test/quantization/core/experimental/test_nonuniform_observer.py

```diff
@@ -6,18 +6,106 @@
 
 class TestNonUniformObserver(unittest.TestCase):
     def test_calculate_qparams(self):
-        t = torch.Tensor()
-        obs = APoTObserver(t, t, t, 0, 0)
+        """
+        Test case 1
+        Assume default parameters for b, k, n (empty observer)
+        """
+        obs1 = APoTObserver()
 
-        with self.assertRaises(NotImplementedError):
-            obs.calculate_qparams()
+        obs1_result = obs1.calculate_qparams(signed=False)
 
-    def test_override_calculate_qparams(self):
-        t = torch.Tensor()
-        obs = APoTObserver(t, t, t, 0, 0)
+        self.assertEqual(obs1_result[0], 0.0)
+        self.assertTrue(torch.equal(obs1_result[1], torch.tensor([])))
+        self.assertTrue(torch.equal(obs1_result[2], torch.tensor([])))
 
-        with self.assertRaises(NotImplementedError):
-            obs._calculate_qparams()
+        """
+        Test case 2
+        APoT paper example: https://arxiv.org/pdf/1909.13144.pdf
+        Assume hardcoded parameters:
+        * b = 4 (total number of bits across all terms)
+        * k = 2 (base bitwidth, i.e. bitwidth of every term)
+        * n = 2 (number of additive terms)
+        * note: b = k * n
+        """
+        obs2 = APoTObserver(max_val=1.0, b=4, k=2)
+        obs2_result = obs2.calculate_qparams(signed=False)
+
+        quantization_levels2_test = torch.tensor([0.0000, 0.0208, 0.0417, 0.0625, 0.0833, 0.1250,
+                                                  0.1667, 0.1875, 0.2500, 0.3333, 0.3750, 0.5000, 0.6667,
+                                                  0.6875, 0.7500, 1.0000])
+        level_indices2_test = torch.tensor([0, 3, 12, 15, 2, 14, 8, 11, 10, 1, 13, 9, 4, 7, 6, 5])
+
+        self.assertEqual(obs2_result[0], (2 / 3))
+        self.assertTrue(torch.allclose(obs2_result[1], quantization_levels2_test, 1e-04, 1e-04))
+        self.assertTrue(torch.equal(obs2_result[2], level_indices2_test))
+
+        """
+        Test case 3
+        Assume hardcoded parameters:
+        * b = 6 (total number of bits across all terms)
+        * k = 2 (base bitwidth, i.e. bitwidth of every term)
+        * n = 3 (number of additive terms)
+        """
+        obs3 = APoTObserver(max_val=1.0, b=6, k=2)
+
+        obs3_result = obs3.calculate_qparams(signed=False)
+
+        quantization_levels3 = obs3_result[1]
+
+        quantization_levels3_test = torch.tensor([0.0000, 0.0022, 0.0045, 0.0067, 0.0089, 0.0112, 0.0134, 0.0156, 0.0179,
```

Comment:
checking the values like this seems problematic. i.e. generally unit tests should verify that something is working through comparison to either a trivial answer or a reference implementation, i.e. its relatively straightforward to see that 2^0+2^-1=1.5 so gamma =2/3 makes sense and if something changes there, its a clear error. Here is a test with a reference comparison: https://www.internalfb.com/code/fbsource/[43fceb50fa4a87c2eded04c8624145a904c1a40c]/fbcode/caffe2/test/quantization/core/test_quantized_op.py?lines=78

on the other hand, i don't know where these hardcoded values came from, they're not trivial to generate and if they were generated by running the code itself, it defeats the purpose to compare with them. 

i'd either 
A) make a reference implementation to compare with 
B) look at higher level things like 'are all the values from 0 to 2**B-1 included in level_indices and does it have any repeats...etc
C) compare with itself when testing certain behaviors, i.e. signed=true vs signed=false should be the same quantization levels, just with both the positive and negative